components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -330,6 +365,252 @@ void AssetDiscoveryManager::MergeDiscoveredEthAssets(
+    std::move(callback).Run(std::move(nfts_so_far));
+    return;
+  }
+
+  // Invalid JSON becomes an empty string after sanitization
+  if (api_request_result.body().empty()) {
+    std::move(callback).Run(std::move(nfts_so_far));
+    return;
+  }
+
+  // optional of a pair of GURL and vector of blockchaintokenptrs
+  absl::optional<std::pair<GURL, std::vector<mojom::BlockchainTokenPtr>>>
+      result = ParseNFTsFromSimpleHash(api_request_result.value_body(), coin);
+  if (!result) {
+    std::move(callback).Run(std::move(nfts_so_far));
+    return;
+  }
+
+  // Add discovered NFTs to nfts_so_far
+  for (auto& token : result.value().second) {
+    nfts_so_far.push_back(std::move(token));
+  }
+
+  // If there is a next page, fetch it
+  if (result.value().first.is_valid()) {
+    auto internal_callback =
+        base::BindOnce(&AssetDiscoveryManager::OnFetchNFTsFromSimpleHash,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(nfts_so_far),
+                       coin, std::move(callback));
+    api_request_helper_->Request("GET", result.value().first, "", "", true,
+                                 std::move(internal_callback));
+    return;
+  }
+
+  // Otherwise, return the nfts_so_far
+  std::move(callback).Run(std::move(nfts_so_far));
+}
+
+absl::optional<std::pair<GURL, std::vector<mojom::BlockchainTokenPtr>>>
+AssetDiscoveryManager::ParseNFTsFromSimpleHash(const base::Value& json_value,
+                                               mojom::CoinType coin) {
+  // Only ETH and SOL NFTs are supported
+  if (!(coin == mojom::CoinType::ETH || coin == mojom::CoinType::SOL)) {
+    return absl::nullopt;
+  }
+
+  if (!json_value.is_dict()) {
+    return absl::nullopt;
+  }
+
+  GURL nextURL;
+  auto* next = json_value.FindStringKey("next");
+  if (next) {
+    nextURL = GURL(*next);
+  }
+
+  const base::Value* nfts = json_value.FindKey("nfts");
+  if (!nfts || !nfts->is_list()) {
+    return absl::nullopt;
+  }
+
+  const base::flat_map<std::string, std::string>& chain_ids =
+      FromSimpleHashChainId();
+  std::vector<mojom::BlockchainTokenPtr> nft_tokens;
+  for (const auto& nft : nfts->GetList()) {
+    auto token = mojom::BlockchainToken::New();
+
+    // skip all tokens with a collection.spam_score > 0
+    auto* collection = nft.FindDictKey("collection");
+    if (!collection) {
+      continue;
+    }
+    absl::optional<int> spam_score = collection->FindIntKey("spam_score");
+    if (!spam_score || *spam_score > 0) {
+      continue;
+    }
+
+    // contract_address (required)
+    auto* contract_address = nft.FindStringKey("contract_address");
+    if (!contract_address) {
+      continue;
+    }
+    token->contract_address = *contract_address;
+
+    // name
+    auto* name = nft.FindStringKey("name");
+    if (name) {
+      token->name = *name;
+    }
+
+    // logo
+    auto* logo = nft.FindStringKey("image_url");
+    if (logo) {
+      token->logo = *logo;
+    }
```

Comment:
Note: I'm just using the SimpleHash URL for the logo.  I think we typically have the frontend fetch the metadata from the chain and use the image url from that as the logo.

Reply:
I think the logo here is the url used to display the icon sized images in certain views.  But in practice I think we've just been using the image_url provided by the metadata for this cc @muliswilliam

Reply:
Ahh ok, so if I'm understanding correctly, in many cases we're just downscaling the actual NFT so that we can display it in various places and the logo would be the ideal image for that where as the actual NFT image would be ideal to display in the viewer. I assume this is necessary for .gif NFTs for example where the logo would be display only, but the NFT viewer would show the actual .gif that moves. Is that a correct way to think about it?

Reply:
That's my understanding, though I'd guess we do that mostly because that's only image URL we have access to in a lot of contexts.

Reply:
Ok, I'm good with the tradeoffs being made here. If we hit a snag during frontend work we can just switch to the `nft.extra_metadata.image_original_url` property too.